### ISSUE TITLE
Unprotected None objects

### ISSUE THREAD

**Issue by sandoro** [svviz/svviz#44]
### svviz-run 11.log

$ svviz --export svviz-run11.svg --format svg --verbose 4 demo
Namespace(_parser=ArgumentParser(prog='svviz',
usage='%(prog)s [options] [demo] [ref breakpoint...] [ref vcf]', 
description='svviz version 1.5.1', version=None, formatter_class=<class 'argparse.RawTextHelpFormatter'>, 
conflict_handler='error', add_help=True), aln_quality=None, aln_score_delta=2, annotations=None, bam=None, 
breakpoints=[], context=0, converter=None, dotplots=False, export='svviz-run11.svg', fasta=None, flanks=False, 
format='svg', include_supplementary=False, lenient=False, max_deletion_size=None, max_multimapping_similarity=0.95, 
max_reads=None, max_size=None, min_mapq=0, no_web=True, open_exported=False, pacbio=False, pair_min_mapq=0, port=None, 
processes=None, ref='demo', sample_reads=None, save_reads=None, save_state=None, skip_cigar=False, summary=None, 
thicker_lines=False, type=None, verbose=4)

\* Sampling reads to calculate Insert Size Distributions *
 Using align distance: None

\* Loading variant *

\* Running for variant 1/1 None *

> Traceback (most recent call last):
>  File "/usr/local/Anaconda/envs_app/svviz/1.5.1/bin/svviz", line 11, 
>    in <module> sys.exit(main())
> 
>  File "/usr/local/Anaconda/envs_app/svviz/1.5.0/lib/python2.7/site-packages/svviz/app.py", line 354, 
>    in main
>    run(sys.argv)
> 
>  File "/usr/local/Anaconda/envs_app/svviz/1.5.0/lib/python2.7/site-packages/svviz/app.py", line 303, 
>    in run
>    debug.printDebugInfo(dataHub)
> 
>  File "/usr/local/Anaconda/envs_app/svviz/1.5.0/lib/python2.7/site-packages/svviz/debug.py", line 12, 
>    in printDebugInfo
>      for part in dataHub.variant.chromParts(allele):
> 
> AttributeError: 'NoneType' object has no attribute 'chromParts'

---
#### 1. The first line shows the command being run
-   --export:     I need to avoid running svviz all the way up to starting the data viewer
               because it doesn't work (must need an ssh tunnel) and I don't want to
               present that as a good approach to our cluster users
-   --format svg: seems most straight-forward format to use
-   --verbose 4:  The arguments on the command line effect the behavior of svviz
  - With 1 argument , "demo", 
    svviz finishes computing and passes to the data viewer which hangs up the
    parent shell of svviz.
  - With 7 arguments , "--export svviz-run11.svg --format svg --verbose 4 demo"
    svviz is diverted into the "debug" module and the error is easier to interpret.
    
    The result is the log file "svviz-run11.log" shown above.
#### 2.   I believe that what is going on is this

```
  *  Because this is a demo session, some internal setup values are initialized to "None".


  *  Some "None" values, in turn, may break code (typically a 'for' loop) when the 
    code assumes that some variable has some attribute that it doesn't actually
    have  because the value of the variable is  "None".


  *  I can see some places in the source code that protect against broken code
    caused by a "None" object, by placing the vulnerable code under an 'if' statement
    like:
```

> ```
>      if variable is not None:
>          USE(variable.attribute)
> ```

```
   *  and I see other places in the source code in which 
     there is no 'if' statement protecting code which accesses an attribute 
     of a variable which could have the "None" object as its value.
     For example, the code reference in the Traceback from "svviz-run11.log" above.
```
#### 3. I don't understand how my proposed explanation can be correct

Am I the only svviz user in the world who has run the svviz demo programs as released?

---

Any insight into this situation would be much appreciated.

Sanford M. Orlow, PhD
Computer Engineer, HPC@NIH
National Institutes of Health, CIT
12 South Dr. 12B/2N207, Bethesda, MD 20892-5680
EMAIL sandor@mail.nih.gov

**Comment by nspies** [comment]
Hrm, I wasn’t intending that “svviz demo” be used with any additional arguments. Sorry there wasn't a better error message — if you check out the latest version from source, it will spit out a more helpful notice.

If you’d like to run the demo but change the arguments, you can do “svviz demo”, type “y” and then hit return to instruct it to download the demo data, and then it will spit out the example command line that it’ll try to run (at this point you can hit ctrl-c to stop it). For example, for the first demo, it’d be:

svviz --type inv -b svviz-examples/example1/son.sorted.bam -b svviz-examples/example1/father.sorted.bam -b svviz-examples/example1/mother.sorted.bam svviz-examples/example1/chr4_short.fa chr4 88847163 88858699

You can copy this command and then edit it. For example, you’d probably want to run this:

svviz --type inv -b svviz-examples/example1/son.sorted.bam -b svviz-examples/example1/father.sorted.bam -b svviz-examples/example1/mother.sorted.bam svviz-examples/example1/chr4_short.fa chr4 88847163 88858699 --export svviz-run11.svg --format svg --verbose 4

Does that make sense?

On Aug 23, 2016, at 7:06 AM, sandoro <notifications@github.com<mailto:notifications@github.com>> wrote:

svviz-run 11.log

$ svviz --export svviz-run11.svg --format svg --verbose 4 demo
Namespace(_parser=ArgumentParser(prog='svviz',
usage='%(prog)s [options] [demo] [ref breakpoint...] [ref vcf]',
description='svviz version 1.5.1', version=None, formatter_class=,
conflict_handler='error', add_help=True), aln_quality=None, aln_score_delta=2, annotations=None, bam=None,
breakpoints=[], context=0, converter=None, dotplots=False, export='svviz-run11.svg', fasta=None, flanks=False,
format='svg', include_supplementary=False, lenient=False, max_deletion_size=None, max_multimapping_similarity=0.95,
max_reads=None, max_size=None, min_mapq=0, no_web=True, open_exported=False, pacbio=False, pair_min_mapq=0, port=None,
processes=None, ref='demo', sample_reads=None, save_reads=None, save_state=None, skip_cigar=False, summary=None,
thicker_lines=False, type=None, verbose=4)
- Sampling reads to calculate Insert Size Distributions *
  Using align distance: None
- Loading variant *
- Running for variant 1/1 None *

Traceback (most recent call last):
File "/usr/local/Anaconda/envs_app/svviz/1.5.1/bin/svviz", line 11,
in sys.exit(main())

File "/usr/local/Anaconda/envs_app/svviz/1.5.0/lib/python2.7/site-packages/svviz/app.py", line 354,
in main
run(sys.argv)

File "/usr/local/Anaconda/envs_app/svviz/1.5.0/lib/python2.7/site-packages/svviz/app.py", line 303,
in run
debug.printDebugInfo(dataHub)

File "/usr/local/Anaconda/envs_app/svviz/1.5.0/lib/python2.7/site-packages/svviz/debug.py", line 12,
in printDebugInfo
for part in dataHub.variant.chromParts(allele):

AttributeError: 'NoneType' object has no attribute 'chromParts'

---

Explanation
1. The first line shows the command being run
-   --export: I need to avoid running svviz all the way up to starting the data viewer
  because it doesn't work (must need an ssh tunnel) and I don't want to
  present that as a good approach to our cluster users
-   --format svg: seems most straight-forward format to use
-   --verbose 4: The arguments on the command line effect the behavior of svviz
  -   With 1 argument , "demo",
    svviz finishes computing and passes to the data viewer which hangs up the
    parent shell of svviz.
  -   With 7 arguments , "--export svviz-run11.svg --format svg --verbose 4 demo"
    svviz is diverted into the "debug" module and the error is easier to interpret.

The result is the log file "svviz-run11.log" shown above.
1. I believe that what is going on is this
   -  Because this is a demo session, some internal setup values are initialized to "None".
   -  Some "None" values, in turn, may break code (typically a 'for' loop) when the
     code assumes that some variable has some attribute that it doesn't actually
     have  because the value of the variable is  "None".
   -  I can see some places in the source code that protect against broken code
     caused by a "None" object, by placing the vulnerable code under an 'if' statement
     like:
     
     if variable is not None:
        USE(variable.attribute)
   -  and I see other places in the source code in which
     there is no 'if' statement protecting code which accesses an attribute
     of a variable which could have the "None" object as its value.
     For example, the code reference in the Traceback from "svviz-run11.log" above.
2. I don't understand how my proposed explanation can be correct

Am I the only svviz user in the world who has run the svviz demo programs as released?

---

Any insight into this situation would be much appreciated.

Sanford M. Orlow, PhD
Computer Engineer, HPC@NIH
National Institutes of Health, CIT
12 South Dr. 12B/2N207, Bethesda, MD 20892-5680
EMAIL sandor@mail.nih.govmailto:sandor@mail.nih.gov

**Comment by nspies** [comment]
closed by 1f3189128ad592609867f5bc533a2c2582b42d91

**Comment by sandoro** [comment]
Noah,

I apologize for not getting back to you about issue #44 for svviz.
Your suggestion about how to run "svviz demo" with additional
svviz command-line options worked perfectly for me and I was able
to complete my testing.

Thanks much for your timely and helpful replies to my questions.

--sandy
HPC@NIH

Sanford M. Orlow, PhD
Computer Engineer, HPC@NIH
National Institutes of Health, CIT
12 South Dr. 12B/2N207, Bethesda, MD 20892-5680
PHONE 301.496.5362 EMAIL sandor@mail.nih.gov

Today (10/17/16) at 01:18 -0400, Noah Spies wrote:

> closed by 1f3189128ad592609867f5bc533a2c2582b42d91

**Comment by nspies** [comment]
Thanks for the update, and glad it worked out!

On Oct 17, 2016, at 7:56 AM, sandoro <notifications@github.com<mailto:notifications@github.com>> wrote:

Noah,

I apologize for not getting back to you about issue #44 for svviz.
Your suggestion about how to run "svviz demo" with additional
svviz command-line options worked perfectly for me and I was able
to complete my testing.

Thanks much for your timely and helpful replies to my questions.

--sandy
HPC@NIH

Sanford M. Orlow, PhD
Computer Engineer, HPC@NIH
National Institutes of Health, CIT
12 South Dr. 12B/2N207, Bethesda, MD 20892-5680
PHONE 301.496.5362 EMAIL sandor@mail.nih.govmailto:sandor@mail.nih.gov

Today (10/17/16) at 01:18 -0400, Noah Spies wrote:
